### ISSUE TITLE
*auto tactics try Hint Resolve hints that don't match their patterns

### ISSUE THREAD

**Issue by jfehrle** [coq/coq#19758]
### Description of the problem

Neither of the `simple apply`s should be tried because their patterns don't match the goal.  Note that the `Extern` hints that have the same patterns are not tried.

I determined that the call to `Bounded_net.lookup` in `Hints.lookup_tacs` is returning the non matching tactics.  Perhaps the pattern isn't added to the discrimination tree correctly for this `Hint Resolve` case.

![image](https://github.com/user-attachments/assets/c28a6c41-7164-45b5-8312-704b6df9199b)

FWIW, in 8.13, it looks like all 4 were tried:

```
Tactic failure: Proof search failed.
Starting resolution with 1 goal(s) under focus and 0 shelved goal(s) in regular mode, unbounded
1: looking for (1 = 0) without backtracking
1.1: (*external*) (idtac "second"; fail) on
(1 = 0) failed with: pattern-matching failed
1.1: (*external*) (idtac "first"; fail) on
(1 = 0) failed with: pattern-matching failed
1.1: simple apply plus_O_n on
(1 = 0) failed with: Cannot unify 0 and 1
1.1: simple apply Nat.add_comm on
(1 = 0) failed with: Cannot unify (?M577 + ?M578 =
                                   ?M578 + ?M577) and 
(1 = 0)
1: no match for (1 = 0), 4 possibilities
```

### Small Rocq / Coq file to reproduce the bug

```coq
From Coq Require Import PeanoNat.
Create HintDb db discriminated.
Hint Resolve Nat.add_comm | 1 : db.
Hint Resolve plus_O_n | 1 : db.
Hint Extern 1 (0 + ?x = ?x) => idtac "first"; fail : db.
Hint Extern 1 (?m + ?n = ?n + ?m) => idtac "second"; fail : db.

Set Typeclasses Debug Verbosity 2.
Goal 1=0.
typeclasses eauto with db nocore.
```

### Version of Rocq / Coq where this bug occurs

8.21 alpha

**Comment by ppedrot** [comment]
Extern hints have a different notion of pattern matching than all others. While the former use a purely syntactic matching, the latter consider patterns up to potential unfolding. As a result, the actual patterns used for the Resolve hints you wrote are basically the catch-all pattern, because the `plus` symbol is transparent. If you want the tactics to treat this constant as an opaque symbol, you have to register it explicitly like this:
```coq
Hint Opaque plus : db.
```

**Comment by jfehrle** [comment]
Is this correct (I want to document this, users should understand this)?
- The code only unfolds items in the goal--not in the hint pattern
- Resolve hints are treated as matches if they contain any unfoldable item (relying on the hint tactic do the matching)?

But I don't see how extern hints are treated differently from resolve hints.  Or the rationale.

**Comment by Janno** [comment]
> The code only unfolds items in the goal--not in the hint pattern

No, the unfolding behavior (ignoring `Hint Extern`) is symmetric. If a constant is transparent, it will be treated as `_` on either side and it will be unfolded by unification on either side. 

It certainly doesn't help that the patterns printed by `Print HintDb` or `Print Hint` are not the ones used for matching. I don't think there currently is a way to print the actual dnet patterns. I have written debug code for this in the past several times but the code was never in a state where it could have been merged. It seems you are doing a lot of work on making the printed output of these commands more readable. Maybe once that's done we can afford to print the dnet pattern and make the output both readable and useful?

I believe there also is no printed representation of the goal term right now. I don't remember if they are technically ever converted to a dnet pattern but even if they are not it would still be correct to print them as such so that people can more easily understand why their TC goal leads to so many hint application attempts.

> Resolve hints are treated as matches if they contain any unfoldable item (relying on the hint tactic do the matching)?

This does not sound like the correct phrasing but I might be too deep into this to say what a good way to phrase it is. On the technical side, hint patterns are translated to dnet patterns, which are trees of labelled nodes. The labels are defined in https://github.com/coq/coq/blob/c7373559ab661b17bd21ea3d0153bb3c82f8d462/tactics/btermdn.ml#L22. They are usually wrapped in some kind of `option` to represent a catch-all node. Children of a node correspond to subterms. `0 + x = x` would be translated to something like this (abusing products and lists to represent the tree):
```
(Some (GRLabel (ConstRef eq),
  [None, (* assuming `plus` is transparent *)
   None (* variables must match everything *)
  ]
)
```
or
```

(Some (GRLabel (IndRef eq),
  [Some (GRLabel  (ConstRef Nat.plus),  (* assuming `plus` is opaque *)
     [Some (GRLabel (ConstructorRef Nat.O)), 
      None (* variables must match everything *)
     ]
   ),
   None (* variables must match everything *)
  ]
)
```

A dnet is basically just a trie that contains many of these trees. The branching in this trie happens on the (optional) label. If `plus` is transparent, it will end up on the same path in the trie as all other hints of the form `f x = x` with `f` transparent. All the hints that match a path are collected in a set at the leaf node of the trie at that path.

I don't know how to explain this except to tell users what is actually happening..

> But I don't see how extern hints are treated differently from resolve hints. Or the rationale.

I know technical reasons for this but only in the sense that I was unable to quickly come up with a way to implement `Hint Extern` that match the same way `Hint Resolve` does. For example, if you have `Hint Extern 0 (C (f ?x)) => tac x` with `f _ := true` and `f` hint transparent, what do you call `tac` with when this hint matches `C true`? Unlike unification variables which can just stay uninstantiated, I don't think the kind of pattern variables that appear in `Hint Extern` are allowed to not have a value.

**Comment by jfehrle** [comment]
> No, the unfolding behavior (ignoring `Hint Extern`) is symmetric.

Right.  I think I experimented with Hint Extern and incorrectly generalized that.

> It seems you are doing a lot of work on making the printed output of [Print HintDb] more readable.

It's an easy starting point and I've always been keen on good user experience.  It helps to have very sensible output while doing experiments.  It's not necessarily the highest priority.  But I do make small improvements as I examine code/functionality.
Furthermore, I'm less likely to break something in a non-obvious way as learn about it.

Regarding dnet form, what info does that provide that you can't get from the pattern expression?  Does that relate to bugs in Coq or things the user can adjust?  Do you need to be a developer to use it?  This may take more code than you think--AFAIK there is no existing printer code for this information.  If it's meant to be shown using notations, we don't have that, either.  And not sure how much documentation it needs--we'd need to explain dnets, among other things, which seem to me to be implementation details.

Aside from the questions above, let's continue discussing after I've finished examining proof search functionality.
(We should keep a list of ideas for improvements for later consideration.)
